File: examples/typescript-content-writer/src/agent.ts
/**
 * Content Writer Agent implementation.
 *
 * This module contains the main agent logic, including:
 * - OpenAI integration for content analysis and generation
 * - Ask-a-Human integration for human decision points
 * - Progress display using chalk and ora
 */

import chalk from "chalk";
import OpenAI from "openai";
import ora from "ora";
import * as fs from "fs";
import * as path from "path";
import { fileURLToPath } from "url";

import {
  AskHumanClient,
  AskHumanOrchestrator,
  AskHumanError,
} from "../../../sdk-typescript/dist/index.js";

import {
  ANALYZE_BRIEF_PROMPT,
  GENERATE_HEADLINES_PROMPT,
  GENERATE_CONTENT_PROMPT,
} from "./prompts.js";

/**
 * Result of analyzing a content brief.
 */
export interface BriefAnalysis {
  contentType: string;
  targetAudience: string;
  keyPoints: string[];
  toneOptions: string[];
}

/**
 * Final generated content with metadata.
 */
export interface ContentResult {
  headline: string;
  tone: string;
  content: string;
  briefAnalysis: BriefAnalysis;
}

/**
 * Agent options for constructor.
 */
export interface ContentWriterAgentOptions {
  openaiApiKey?: string;
  askHumanAgentId?: string;
  askHumanBaseUrl?: string;
}

/**
 * Agent that writes content with human-in-the-loop decisions.
 *
 * This agent demonstrates the async human-in-the-loop pattern:
 * 1. Analyze brief with LLM
 * 2. Ask humans for tone preference
 * 3. Generate headline options with LLM
 * 4. Ask humans for headline preference
 * 5. Generate final content with LLM
 */
export class ContentWriterAgent {
  private openai: OpenAI;
  private askHuman: AskHumanClient;
  private orchestrator: AskHumanOrchestrator;

  /**
   * Initialize the agent.
   */
  constructor(options?: ContentWriterAgentOptions) {
    // Initialize OpenAI client
    const apiKey = options?.openaiApiKey ?? this.getOpenAIKey();
    this.openai = new OpenAI({ apiKey });

    // Initialize Ask-a-Human client
    this.askHuman = new AskHumanClient({
      baseUrl: options?.askHumanBaseUrl,
      agentId: options?.askHumanAgentId ?? "content-writer-agent",
    });

    this.orchestrator = new AskHumanOrchestrator(this.askHuman, {
      pollInterval: 10000, // Check every 10 seconds
      maxBackoff: 60000, // Max 1 minute between checks
    });
  }

  /**
   * Get OpenAI API key from env var or secrets file.
   */
  private getOpenAIKey(): string {
    // Try environment variable first
    const envKey = process.env["OPENAI_API_KEY"];
    if (envKey) {
      return envKey;
    }

    // Try secrets file
    const __filename = fileURLToPath(import.meta.url);
    const __dirname = path.dirname(__filename);
    const secretsPath = path.join(__dirname, "..", "..", "..", "..", "secrets", "openai-api-key.txt");

    if (fs.existsSync(secretsPath)) {
      return fs.readFileSync(secretsPath, "utf-8").trim();
    }

    throw new Error(
      "OpenAI API key not found. Set OPENAI_API_KEY env var or create secrets/openai-api-key.txt"
    );
  }

  /**
   * Call OpenAI with a prompt.
   */
  private async callLLM(prompt: string, parseJson: boolean = false): Promise<string | Record<string, unknown>> {
    const response = await this.openai.chat.completions.create({
      model: "gpt-4o-mini",
      messages: [{ role: "user", content: prompt }],
      temperature: 0.7,
    });

    let text = response.choices[0]?.message?.content ?? "";

    if (parseJson) {
      // Extract JSON from response (handle markdown code blocks)
      if (text.includes("```json")) {
        text = text.split("```json")[1]?.split("```")[0] ?? text;
      } else if (text.includes("```")) {
        text = text.split("```")[1]?.split("```")[0] ?? text;
      }
      return JSON.parse(text) as Record<string, unknown>;
    }

    return text;
  }

  /**
   * Analyze a content brief using OpenAI.
   */
  async analyzeBrief(brief: string): Promise<BriefAnalysis> {
    console.log(chalk.blue("\n📋 Analyzing your brief..."));

    const prompt = ANALYZE_BRIEF_PROMPT.replace("{brief}", brief);
    const result = await this.callLLM(prompt, true);

    if (typeof result === "string") {
      throw new Error(`Expected JSON response, got: ${result}`);
    }

    const analysis: BriefAnalysis = {
      contentType: (result["content_type"] as string) ?? "article",
      targetAudience: (result["target_audience"] as string) ?? "general audience",
      keyPoints: (result["key_points"] as string[]) ?? [],
      toneOptions: (result["tone_options"] as string[]) ?? [
        "Formal and professional",
        "Casual and friendly",
        "Informative and neutral",
        "Engaging and persuasive",
      ],
    };

    console.log(
      chalk.green("✓") + ` Brief analyzed: ${analysis.contentType} for ${analysis.targetAudience}`
    );

    return analysis;
  }

  /**
   * Ask humans to choose the tone/style.
   */
  async askHumansForTone(analysis: BriefAnalysis): Promise<string> {
    console.log(chalk.blue("\n🧑 Asking humans about writing tone..."));

    try {
      // Submit the question
      const submission = await this.orchestrator.submit({
        prompt: `We're writing a ${analysis.contentType} for ${analysis.targetAudience}. Which tone/style would be most effective?`,
        type: "multiple_choice",
        options: analysis.toneOptions,
        audience: ["creative", "product"],
        minResponses: 3,
        timeoutSeconds: 600, // 10 minutes
      });

      console.log(`   Question ID: ${submission.questionId}`);

      // Wait for responses with spinner
      const spinner = ora("Waiting for human responses...").start();

      const responses = await this.orchestrator.awaitResponses([submission.questionId], {
        minResponses: 1, // Accept at least 1 response
        timeout: 300000, // Wait up to 5 minutes
      });

      spinner.stop();

      const question = responses[submission.questionId];
      if (!question) {
        throw new Error("No response received");
      }

      // Count votes for each option
      let selectedTone: string;

      if (question.summary) {
        // Find the winning option
        const entries = Object.entries(question.summary);
        if (entries.length > 0) {
          const winner = entries.reduce((a, b) => (a[1] > b[1] ? a : b));
          selectedTone = winner[0];
        } else {
          selectedTone = analysis.toneOptions[0] ?? "Professional";
        }
      } else if (question.responses.length > 0) {
        // Fall back to first response
        const firstResponse = question.responses[0];
        if (firstResponse?.selectedOption !== undefined) {
          selectedTone = analysis.toneOptions[firstResponse.selectedOption] ?? analysis.toneOptions[0] ?? "Professional";
        } else {
          selectedTone = analysis.toneOptions[0] ?? "Professional";
        }
      } else {
        // No responses, use default
        console.log(chalk.yellow("⚠") + " No human responses received, using default tone");
        selectedTone = analysis.toneOptions[0] ?? "Professional";
      }

      console.log(chalk.green("✓") + ` Tone preference received: "${selectedTone}"`);
      return selectedTone;
    } catch (error) {
      if (error instanceof AskHumanError) {
        console.log(chalk.yellow("⚠") + ` Ask-a-Human error: ${error.message}`);
        console.log("   Using default tone");
        return analysis.toneOptions[0] ?? "Professional";
      }
      throw error;
    }
  }

  /**
   * Generate headline options using OpenAI.
   */
  async generateHeadlines(brief: string, analysis: BriefAnalysis, tone: string): Promise<string[]> {
    console.log(chalk.blue("\n💡 Generating headline options..."));

    const prompt = GENERATE_HEADLINES_PROMPT
      .replace("{brief}", brief)
      .replace("{tone}", tone)
      .replace("{contentType}", analysis.contentType)
      .replace("{targetAudience}", analysis.targetAudience);

    const result = await this.callLLM(prompt, true);

    if (typeof result === "string") {
      throw new Error(`Expected JSON response, got: ${result}`);
    }

    const headlines = (result["headlines"] as string[]) ?? [
      "Compelling Headline Option 1",
      "Engaging Headline Option 2",
      "Powerful Headline Option 3",
      "Creative Headline Option 4",
    ];

    console.log(chalk.green("✓") + ` Generated ${headlines.length} headline options`);
    return headlines;
  }

  /**
   * Ask humans to choose the headline.
   */
  async askHumansForHeadline(headlines: string[]): Promise<string> {
    console.log(chalk.blue("\n🧑 Asking humans about headline preference..."));

    try {
      // Submit the question
      const submission = await this.orchestrator.submit({
        prompt: "Which headline is most compelling and would make you want to read more?",
        type: "multiple_choice",
        options: headlines,
        audience: ["creative", "product"],
        minResponses: 3,
        timeoutSeconds: 600,
      });

      console.log(`   Question ID: ${submission.questionId}`);

      // Wait for responses
      const spinner = ora("Waiting for human responses...").start();

      const responses = await this.orchestrator.awaitResponses([submission.questionId], {
        minResponses: 1,
        timeout: 300000,
      });

      spinner.stop();

      const question = responses[submission.questionId];
      if (!question) {
        throw new Error("No response received");
      }

      // Count votes
      let selectedHeadline: string;

      if (question.summary) {
        const entries = Object.entries(question.summary);
        if (entries.length > 0) {
          const winner = entries.reduce((a, b) => (a[1] > b[1] ? a : b));
          selectedHeadline = winner[0];
        } else {
          selectedHeadline = headlines[0] ?? "Untitled";
        }
      } else if (question.responses.length > 0) {
        const firstResponse = question.responses[0];
        if (firstResponse?.selectedOption !== undefined) {
          selectedHeadline = headlines[firstResponse.selectedOption] ?? headlines[0] ?? "Untitled";
        } else {
          selectedHeadline = headlines[0] ?? "Untitled";
        }
      } else {
        console.log(chalk.yellow("⚠") + " No human responses received, using first headline");
        selectedHeadline = headlines[0] ?? "Untitled";
      }

      console.log(chalk.green("✓") + ` Headline selected: "${selectedHeadline}"`);
      return selectedHeadline;
    } catch (error) {
      if (error instanceof AskHumanError) {
        console.log(chalk.yellow("⚠") + ` Ask-a-Human error: ${error.message}`);
        console.log("   Using first headline option");
        return headlines[0] ?? "Untitled";
      }
      throw error;
    }
  }

  /**
   * Generate the final content using OpenAI.
   */
  async generateContent(
    brief: string,
    analysis: BriefAnalysis,
    tone: string,
    headline: string
  ): Promise<string> {
    console.log(chalk.blue("\n📝 Generating final content..."));

    const keyPointsFormatted = analysis.keyPoints.map((point) => `- ${point}`).join("\n");

    const prompt = GENERATE_CONTENT_PROMPT
      .replace("{brief}", brief)
      .replace("{contentType}", analysis.contentType)
      .replace("{targetAudience}", analysis.targetAudience)
      .replace("{tone}", tone)
      .replace("{headline}", headline)
      .replace("{keyPoints}", keyPointsFormatted);

    const content = await this.callLLM(prompt, false);

    if (typeof content !== "string") {
      throw new Error("Expected string response");
    }

    console.log(chalk.green("✓") + " Content generated");
    return content;
  }

  /**
   * Run the full content writing workflow.
   */
  async run(brief: string): Promise<ContentResult> {
    // Step 1: Analyze the brief
    const analysis = await this.analyzeBrief(brief);

    // Step 2: Ask humans for tone (HUMAN DECISION POINT 1)
    const tone = await this.askHumansForTone(analysis);

    // Step 3: Generate headline options
    const headlines = await this.generateHeadlines(brief, analysis, tone);

    // Step 4: Ask humans for headline (HUMAN DECISION POINT 2)
    const headline = await this.askHumansForHeadline(headlines);

    // Step 5: Generate final content
    const content = await this.generateContent(brief, analysis, tone, headline);

    return {
      headline,
      tone,
      content,
      briefAnalysis: analysis,
    };
  }
}

File: examples/content-writer-agent/agent.py
"""Content Writer Agent implementation.

This module contains the main agent logic, including:
- OpenAI integration for content analysis and generation
- Ask-a-Human integration for human decision points
- Progress display using rich
"""

from __future__ import annotations

import json
import os
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from openai import OpenAI
from rich.console import Console
from rich.live import Live
from rich.panel import Panel
from rich.progress import Progress, SpinnerColumn, TextColumn
from rich.table import Table

from ask_a_human import AskHumanClient, AskHumanOrchestrator
from ask_a_human.exceptions import AskHumanError

from prompts import ANALYZE_BRIEF_PROMPT, GENERATE_CONTENT_PROMPT, GENERATE_HEADLINES_PROMPT


@dataclass
class BriefAnalysis:
    """Result of analyzing a content brief."""

    content_type: str
    target_audience: str
    key_points: list[str]
    tone_options: list[str]


@dataclass
class ContentResult:
    """Final generated content with metadata."""

    headline: str
    tone: str
    content: str
    brief_analysis: BriefAnalysis


class ContentWriterAgent:
    """Agent that writes content with human-in-the-loop decisions.

    This agent demonstrates the async human-in-the-loop pattern:
    1. Analyze brief with LLM
    2. Ask humans for tone preference
    3. Generate headline options with LLM
    4. Ask humans for headline preference
    5. Generate final content with LLM
    """

    def __init__(
        self,
        console: Console | None = None,
        openai_api_key: str | None = None,
        ask_human_agent_id: str | None = None,
        ask_human_base_url: str | None = None,
    ) -> None:
        """Initialize the agent.

        Args:
            console: Rich console for output. Creates one if not provided.
            openai_api_key: OpenAI API key. Falls back to env var or secrets file.
            ask_human_agent_id: Agent ID for Ask-a-Human. Falls back to env var.
            ask_human_base_url: Base URL for Ask-a-Human API. Falls back to env var.
        """
        self.console = console or Console()

        # Initialize OpenAI client
        api_key = openai_api_key or self._get_openai_key()
        self.openai = OpenAI(api_key=api_key)

        # Initialize Ask-a-Human client
        self.ask_human = AskHumanClient(
            base_url=ask_human_base_url,
            agent_id=ask_human_agent_id or "content-writer-agent",
        )
        self.orchestrator = AskHumanOrchestrator(
            self.ask_human,
            poll_interval=10.0,  # Check every 10 seconds
            max_backoff=60.0,  # Max 1 minute between checks
        )

    def _get_openai_key(self) -> str:
        """Get OpenAI API key from env var or secrets file."""
        # Try environment variable first
        if key := os.environ.get("OPENAI_API_KEY"):
            return key

        # Try secrets file
        secrets_path = Path(__file__).parent.parent.parent / "secrets" / "openai-api-key.txt"
        if secrets_path.exists():
            return secrets_path.read_text().strip()

        raise ValueError(
            "OpenAI API key not found. Set OPENAI_API_KEY env var "
            "or create secrets/openai-api-key.txt"
        )

    def _call_llm(self, prompt: str, parse_json: bool = False) -> str | dict[str, Any]:
        """Call OpenAI with a prompt.

        Args:
            prompt: The prompt to send.
            parse_json: Whether to parse the response as JSON.

        Returns:
            The response text, or parsed JSON if parse_json=True.
        """
        response = self.openai.chat.completions.create(
            model="gpt-4o-mini",
            messages=[{"role": "user", "content": prompt}],
            temperature=0.7,
        )

        text = response.choices[0].message.content or ""

        if parse_json:
            # Extract JSON from response (handle markdown code blocks)
            if "```json" in text:
                text = text.split("```json")[1].split("```")[0]
            elif "```" in text:
                text = text.split("```")[1].split("```")[0]
            return json.loads(text)

        return text

    def analyze_brief(self, brief: str) -> BriefAnalysis:
        """Analyze a content brief using OpenAI.

        Args:
            brief: The content brief from the user.

        Returns:
            BriefAnalysis with content type, audience, key points, and tone options.
        """
        self.console.print("\n[bold blue]📋 Analyzing your brief...[/bold blue]")

        prompt = ANALYZE_BRIEF_PROMPT.format(brief=brief)
        result = self._call_llm(prompt, parse_json=True)

        if isinstance(result, str):
            raise ValueError(f"Expected JSON response, got: {result}")

        analysis = BriefAnalysis(
            content_type=result.get("content_type", "article"),
            target_audience=result.get("target_audience", "general audience"),
            key_points=result.get("key_points", []),
            tone_options=result.get("tone_options", [
                "Formal and professional",
                "Casual and friendly",
                "Informative and neutral",
                "Engaging and persuasive",
            ]),
        )

        self.console.print(
            f"[green]✓[/green] Brief analyzed: {analysis.content_type} "
            f"for {analysis.target_audience}"
        )

        return analysis

    def ask_humans_for_tone(self, analysis: BriefAnalysis) -> str:
        """Ask humans to choose the tone/style.

        Args:
            analysis: The brief analysis with tone options.

        Returns:
            The selected tone.
        """
        self.console.print("\n[bold blue]🧑 Asking humans about writing tone...[/bold blue]")

        try:
            # Submit the question
            submission = self.orchestrator.submit(
                prompt=(
                    f"We're writing a {analysis.content_type} for {analysis.target_audience}. "
                    f"Which tone/style would be most effective?"
                ),
                type="multiple_choice",
                options=analysis.tone_options,
                audience=["creative", "product"],
                min_responses=3,
                timeout_seconds=600,  # 10 minutes
            )

            self.console.print(f"   Question ID: {submission.question_id}")

            # Wait for responses with progress display
            with Progress(
                SpinnerColumn(),
                TextColumn("[progress.description]{task.description}"),
                console=self.console,
            ) as progress:
                task = progress.add_task("Waiting for human responses...", total=None)

                responses = self.orchestrator.await_responses(
                    [submission.question_id],
                    min_responses=1,  # Accept at least 1 response
                    timeout=300,  # Wait up to 5 minutes
                )

            question = responses[submission.question_id]

            # Count votes for each option
            if question.summary:
                # Find the winning option
                winner = max(question.summary.items(), key=lambda x: x[1])
                selected_tone = winner[0]
            elif question.responses:
                # Fall back to first response
                first_response = question.responses[0]
                if first_response.selected_option is not None:
                    selected_tone = analysis.tone_options[first_response.selected_option]
                else:
                    selected_tone = analysis.tone_options[0]
            else:
                # No responses, use default
                self.console.print(
                    "[yellow]⚠[/yellow] No human responses received, using default tone"
                )
                selected_tone = analysis.tone_options[0]

            self.console.print(f'[green]✓[/green] Tone preference received: "{selected_tone}"')
            return selected_tone

        except AskHumanError as e:
            self.console.print(f"[yellow]⚠[/yellow] Ask-a-Human error: {e}")
            self.console.print("   Using default tone")
            return analysis.tone_options[0]

    def generate_headlines(self, brief: str, analysis: BriefAnalysis, tone: str) -> list[str]:
        """Generate headline options using OpenAI.

        Args:
            brief: The original content brief.
            analysis: The brief analysis.
            tone: The selected tone.

        Returns:
            List of headline options.
        """
        self.console.print("\n[bold blue]💡 Generating headline options...[/bold blue]")

        prompt = GENERATE_HEADLINES_PROMPT.format(
            brief=brief,
            tone=tone,
            content_type=analysis.content_type,
            target_audience=analysis.target_audience,
        )

        result = self._call_llm(prompt, parse_json=True)

        if isinstance(result, str):
            raise ValueError(f"Expected JSON response, got: {result}")

        headlines = result.get("headlines", [
            "Compelling Headline Option 1",
            "Engaging Headline Option 2",
            "Powerful Headline Option 3",
            "Creative Headline Option 4",
        ])

        self.console.print(f"[green]✓[/green] Generated {len(headlines)} headline options")
        return headlines

    def ask_humans_for_headline(self, headlines: list[str]) -> str:
        """Ask humans to choose the headline.

        Args:
            headlines: List of headline options.

        Returns:
            The selected headline.
        """
        self.console.print("\n[bold blue]🧑 Asking humans about headline preference...[/bold blue]")

        try:
            # Submit the question
            submission = self.orchestrator.submit(
                prompt="Which headline is most compelling and would make you want to read more?",
                type="multiple_choice",
                options=headlines,
                audience=["creative", "product"],
                min_responses=3,
                timeout_seconds=600,
            )

            self.console.print(f"   Question ID: {submission.question_id}")

            # Wait for responses
            with Progress(
                SpinnerColumn(),
                TextColumn("[progress.description]{task.description}"),
                console=self.console,
            ) as progress:
                task = progress.add_task("Waiting for human responses...", total=None)

                responses = self.orchestrator.await_responses(
                    [submission.question_id],
                    min_responses=1,
                    timeout=300,
                )

            question = responses[submission.question_id]

            # Count votes
            if question.summary:
                winner = max(question.summary.items(), key=lambda x: x[1])
                selected_headline = winner[0]
            elif question.responses:
                first_response = question.responses[0]
                if first_response.selected_option is not None:
                    selected_headline = headlines[first_response.selected_option]
                else:
                    selected_headline = headlines[0]
            else:
                self.console.print(
                    "[yellow]⚠[/yellow] No human responses received, using first headline"
                )
                selected_headline = headlines[0]

            self.console.print(f'[green]✓[/green] Headline selected: "{selected_headline}"')
            return selected_headline

        except AskHumanError as e:
            self.console.print(f"[yellow]⚠[/yellow] Ask-a-Human error: {e}")
            self.console.print("   Using first headline option")
            return headlines[0]

    def generate_content(
        self, brief: str, analysis: BriefAnalysis, tone: str, headline: str
    ) -> str:
        """Generate the final content using OpenAI.

        Args:
            brief: The original content brief.
            analysis: The brief analysis.
            tone: The selected tone.
            headline: The selected headline.

        Returns:
            The generated content.
        """
        self.console.print("\n[bold blue]📝 Generating final content...[/bold blue]")

        prompt = GENERATE_CONTENT_PROMPT.format(
            brief=brief,
            content_type=analysis.content_type,
            target_audience=analysis.target_audience,
            tone=tone,
            headline=headline,
            key_points="\n".join(f"- {point}" for point in analysis.key_points),
        )

        content = self._call_llm(prompt, parse_json=False)

        if not isinstance(content, str):
            content = str(content)

        self.console.print("[green]✓[/green] Content generated")
        return content

    def run(self, brief: str) -> ContentResult:
        """Run the full content writing workflow.

        Args:
            brief: The content brief from the user.

        Returns:
            ContentResult with the generated content and metadata.
        """
        # Step 1: Analyze the brief
        analysis = self.analyze_brief(brief)

        # Step 2: Ask humans for tone (HUMAN DECISION POINT 1)
        tone = self.ask_humans_for_tone(analysis)

        # Step 3: Generate headline options
        headlines = self.generate_headlines(brief, analysis, tone)

        # Step 4: Ask humans for headline (HUMAN DECISION POINT 2)
        headline = self.ask_humans_for_headline(headlines)

        # Step 5: Generate final content
        content = self.generate_content(brief, analysis, tone, headline)

        return ContentResult(
            headline=headline,
            tone=tone,
            content=content,
            brief_analysis=analysis,
        )

    def close(self) -> None:
        """Clean up resources."""
        self.ask_human.close()
